File: 0091-decode-ways/0091-decode-ways.ts
function numDecodings(s: string): number {
    
 let n=s.length;
    let dp=new Array(n+1).fill(0);
    dp[n]=1;
    dp[n-1]=s[n-1]=="0"?0:1;
    for(let i=n-2;i>=0;i--){
        if(s[i]=="0"){
            continue;
        }
        dp[i]=dp[i+1]+(s[i]=="1"||s[i]=="2"&&s[i+1]<="6"?dp[i+2]:0);
    }
    console.log(dp);
    return dp[0];
};

File: 0091-decode-ways/0091-decode-ways.py
class Solution:
    def numDecodings(self, s: str) -> int:
        dfs = {}

        def helper(s):
            if s in dfs:
                return dfs[s]
            if not s:
                return 1
            if s[0] == '0':
                return 0
            if len(s) == 1:
                return 1
            res = helper(s[1:])
            if int(s[:2]) <= 26:
                res += helper(s[2:])
            dfs[s] = res
            return res

        return helper(s)
        